device: Wiegand badge reader support

brianredbeard opened the pull request:
This commit provides support for [Wiegand][1] style badge readers.  It has
been tested with a number of EM4100 style readers and has been used in
production for a number of months without issue.

The wiring for a badge reader requires two data GPIO pins as input.
These pins within the Wiegand protocol are "D0" and "D1" respectively.
Normally, these pins are pulled high (generally to 5v).  They perform
differential signalling through pulling D0 low to represent a "0" and
pulling D1 low to represent a "1".

This introduces some challenges as most of the connectors have only a
single raw GPIO input.  In practice, this has been used with J5
consuming the B5_LOGIC and B6_LOGIC pins (pin 13 and 7 respectively on
version 1.0 of the Makerspace Auth board).

Within the code this bitstream of "1"s and "0"s is represented as a
string so as to avoid additional dependencies as well as to not worry
about specific card formats.  For our purposes, this module should
pedantically output the hexadecimal value and leave card ids, facility
codes, and other format specific nuances to other components of
software.

[1]: https://en.wikipedia.org/wiki/Wiegand_interface

brianredbeard committed:
device: Wiegand badge reader support

This commit provides support for [Wiegand][1] style badge readers.  It has
been tested with a number of EM4100 style readers and has been used in
production for a number of months without issue.

The wiring for a badge reader requires two data GPIO pins as input.
These pins within the Wiegand protocol are "D0" and "D1" respectively.
Normally, these pins are pulled high (generally to 5v).  They perform
differential signalling through pulling D0 low to represent a "0" and
pulling D1 low to represent a "1".

This introduces some challenges as most of the connectors have only a
single raw GPIO input.  In practice, this has been used with J5
consuming the B5_LOGIC and B6_LOGIC pins (pin 13 and 7 respectively on
version 1.0 of the Makerspace Auth board).

Within the code this bitstream of "1"s and "0"s is represented as a
string so as to avoid additional dependencies as well as to not worry
about specific card formats.  For our purposes, this module should
pedantically output the hexadecimal value and leave card ids, facility
codes, and other format specific nuances to other components of
software.

[1]: https://en.wikipedia.org/wiki/Wiegand_interface

brianredbeard commented:
Affects #7 

Once this has been validated, I will create a separate PR with documentation for the updated version of 

https://github.com/brianredbeard/makerspace-auth/blob/7a3651881a2b50646ae64f6251ddcb6556826bfc/software/doorauth.py

cc @thatch

brianredbeard committed:
bug: missed changes to api.py

After talking to Tim about the structure of the module, he requested
that BaseWiegandPinThread be added to api.py and then pull in the
components from there.

thatch commented on software/authbox/api.py:
```diff
@@ -137,6 +138,21 @@ def __init__(self, event_queue, config_name, input_pin, output_pin, initial_outp
 
 
+class BaseWiegandPinThread(BaseDerivedThread):
+  def __init__(self, event_queue, config_name, d0_pin, d1_pin, initial_output=GPIO.LOW):
```
`initial_output` is unused, and can be removed.

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+# Copyright 2018 Ace Monster Toys. All Rights Reserved.
```
-2019?

brianredbeard replied:
This is correct as the code first started in 2018, thus making 2018 the beginning of the expressed time of copyright (and correctly set the date for expiration releasing it into the commons).

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+    [pins]
+    name = WiegandGPIOReader:3:5
+
+  where 3 is the D0 pin (physical numbering), and 5 is the D1 pin (also 
```
I know these are kind of minimal examples, but could you make it use the pins from J5?  When you add more docs, this should link to there.

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+  # simplify its processing.  It was initially implemented using the
+  # "bitstream" package, but this then had a dependency on a C compiler and
+  # NumPy
+  bits = ''
```
I would rather see bits and timeout treated as instance variables (set in `__init__`) and DEFAULT_TIMEOUT_IN_MS as a module constant.  In your experimenting, how many ms apart were the bits (that's worth documenting in a comment)...

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+  def_timeout_in_ms = 15
+  timeout = def_timeout_in_ms
+
+  def __init__( self, event_queue, config_name, d0_pin, d1_pin, on_scan=None):
```
you have an extra space before `self`

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+      self.event_queue.put((self._on_scan, self))
+
+  def decode(self, channel):
+      if channel == self.d0_pin:
```
Worth noting somewhere that these callbacks are single-threaded from RPi.GPIO.  The run_inner is in a separate thread though, so if there are observed races then this could switch to a threadsafe queue.

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+    the stream, this will actually hold in the timeout state.
+
+    Returns:
+      badge value as string
```
...of 0's and 1's

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+        #print('{0:30b}'.format(int(self.bits[-25:-1], 2)))
+        b = self.bits
+        self.bits = ''
+        # change to start=1 to avoid this calculation and the one in
```
I agree with the comment, `b[1:-1]` would be cleaner but how did you determine those two bits should be stripped?

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+        GPIO.add_event_detect(self.d1_pin, GPIO.FALLING, callback=self.decode)
+
+
+  def _callback(self):
```
This function isn't used.

thatch commented on software/authbox/badgereader_wiegand_gpio.py:
```diff
@@ -0,0 +1,107 @@
+"""
+
+from authbox.api import BaseWiegandPinThread, GPIO
+import time
```
style: stdlib imports typically come first (https://www.python.org/dev/peps/pep-0008/#imports)

thatch committed:
Add tests and switch to queue for badgereader_wiegand.

The queue actually makes the testing logic way easier.

thatch commented:
Mostly style suggestions above.  Also take a look at https://github.com/thatch/makerspace-auth/commit/f100d31e227aa26cbabc8dfa77ad170669ea8010 which adds tests and reworks it to use a queue.  (You're welcome to squash including that and force-push.)

I would suggest either making the bit stripping customizable, or (as in the commit above) making a binary string the interface instead of a hex string.

brianredbeard committed:
FIX: Making requested changes to #28

In pr/#28 @thatch requested a number of changes made.  Some of which
were handled in the previous cherry pick the rest of which are added in
this commit.

thatch commented:
@precision of course I'm okay with that being included.  Ready to merge.

precision commented:
"The bad news is that it appears that one or more commits were authored or co-authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request."

thatch commented:
@precision yes, I was responding to that message in my previous comment "I'm ok with that being included."  I'm the other author, and the code is what I offered on June 5.

brianredbeard commented:
Would it be easier for me to just pave this, re-do the commits, and re-submit?

precision commented:
It'd probably be faster than me waiting for this silly approval, yes.
